### ISSUE TITLE
Bad error message when Signing URLs without a service account email

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp#11740]
The storage client can created signed URLs.  These are URLs that can access GCS resources via a *very* specific request without the need for additional authentication.  This involves creating a the request in a canonical form, hashing it, and signing the request and the hash.

When the credentials are obtained from a service account key file, the client library can sign the canonical request plus hash (it is just treated as a blob at the signing stage) locally, using the private key from the key file.

When the credentials are not a service account key file, the client library can use [IAM signBlob](https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/signBlob) to sign the blob.

In the latter case it needs the service account email of the request to sign. For some credential types there is a default email associated with the credentials.  But sometimes there is none, or the value cannot be retrieved, for example, the metadata service may be unresponsive.

Without a default value for the service account email, the client library sends a request to `IAM signBlob` with an empty email address.  This results in a 404 error without additional details.

This is a case where the client library should not rely on the service to return an error.  We should make an exception to our [ADR](https://github.com/googleapis/google-cloud-cpp/blob/main/doc/adr/2018-06-19-do-not-duplicate-server-side-validation.md) and return a more meaningful error.  An empty string here will never succeed, it is a common error, and the returned error lacks enough information to troubleshoot effectively.

**Comment by jhanley-com** [comment]
This issue is from https://stackoverflow.com/questions/76345053/how-to-set-the-project-id-for-c-storage-client

User ADC: If the environment variable `GOOGLE_CLOUD_PROJECT` is set, the `CreateV4SignedUrl` works without specifying the SigningAccount parameter. If the environment variable is not set, then the SigningAccount parameter is required.

However, the Project ID should be able to be set via the `gcs::Client()` option `gcs::ProjectIdOption>(project_id)`, but that seems to be unimplemented.

**Comment by coryan** [comment]
> This issue is from https://stackoverflow.com/questions/76345053/how-to-set-the-project-id-for-c-storage-client

Thanks for the summary. That is very useful when trying to repro the issue.

> 
> User ADC: If the environment variable `GOOGLE_CLOUD_PROJECT` is set, the `CreateV4SignedUrl` works without specifying the SigningAccount parameter. If the environment variable is not set, then the SigningAccount parameter is required.

That is not what I see.  Which only means I must be doing something different than you are.  Maybe you can read my comments below and tell me what I did differently?  I am using User ADC:

```
unset GOOGLE_APPLICATION_CREDENTIALS
grep type ~/.config/gcloud/application_default_credentials.json 
  "type": "authorized_user"
```

I modified the example to skip the `SigningAccount` parameter:

```
git diff
diff --git a/google/cloud/storage/examples/storage_signed_url_v4_samples.cc b/google/cloud/storage/examples/storage_signed_url_v4_samples.cc
index 81c6a86c25..6486070952 100644
--- a/google/cloud/storage/examples/storage_signed_url_v4_samples.cc
+++ b/google/cloud/storage/examples/storage_signed_url_v4_samples.cc
@@ -25,11 +25,10 @@ void CreateGetSignedUrlV4(google::cloud::storage::Client client,
   namespace gcs = ::google::cloud::storage;
   using ::google::cloud::StatusOr;
   [](gcs::Client client, std::string const& bucket_name,
-     std::string const& object_name, std::string const& signing_account) {
+     std::string const& object_name, std::string const& ) {
     StatusOr<std::string> signed_url = client.CreateV4SignedUrl(
         "GET", std::move(bucket_name), std::move(object_name),
-        gcs::SignedUrlDuration(std::chrono::minutes(15)),
-        gcs::SigningAccount(signing_account));
+        gcs::SignedUrlDuration(std::chrono::minutes(15)));
 
     if (!signed_url) throw std::move(signed_url).status();
     std::cout << "The signed url is: " << *signed_url << "\n\n"
```

When I run this example it fails, even if the environment variable is set:

```
$ GOOGLE_CLOUD_PROJECT=coryan-test  .build/google/cloud/storage/examples/storage_signed_url_v4_samples create-get-signed-url-v4 coryan-test-bucket hello.txt ""
google::cloud::Status thrown: NOT_FOUND: Permanent error SignBlob: Received HTTP status code: 404
terminate called after throwing an instance of 'google::cloud::v2_11::Status'
```

With additional logging you can see that this is an error in the `IAM signBlob` RPC:

```
$ GOOGLE_CLOUD_PROJECT=coryan-test CLOUD_STORAGE_ENABLE_TRACING=raw-client,http GOOGLE_CLOUD_CPP_ENABLE_CLOG=Y .build/google/cloud/storage/examples/storage_signed_url_v4_samples create-get-signed-url-v4 coryan-test-bucket hello.txt ""
... ... ...
>> curl(Send Header): POST /v1/projects/-/serviceAccounts/:signBlob HTTP/1.1
... ... ...
<< curl(Recv Header): HTTP/1.1 404 Not Found
```

The same program with a service account key works fine.

> However, the Project ID should be able to be set via the `gcs::Client()` option `gcs::ProjectIdOption>(project_id)`, but that seems to be implemented but never processed later in the code.

I agree that the behavior around `ProjectIdOption` and `GOOGLE_CLOUD_PROJECT` is inconsistent and we should fix the inconsistencies (see ##11742). I am not sure it would have any effect on *this* bug.
